Add functions for serializing keys as text

rvl opened the pull request:
Relates to #143.

# Overview

To store XPrv and XPub in a database column or text file, a hex-encoded string is suitable.

KtorZ commented on src/Cardano/Wallet/API/V1/Types.hs:
```diff
@@ -2409,6 +2410,28 @@ instance FromJSON ErrNotEnoughMoney where
+        (ErrUtxoNotEnoughFragmentedBy missingUtxos) -> object
+            [ "msg"          .= String "Utxo is not enough fragmented to handle the number of outputs of this transaction. see /api/v1/wallets/{walletId}/statistics/utxos for more information"
+            , "missingUtxos" .= missingUtxos
+            ]
```
API errors are kinda... complicated at the moment but we do want to derive most things generically. I'd suggest to have:

```hs
data ErrUtxoNotEnoughFragmented = ErrUtxoNotEnoughFragmented
    { missingUtxos :: Int
    , msg :: Text
    }
```

And, provide a corresponding message as text in the `ReifyError`. This way, we can simply derive the ToJSON and FromJSON instance generically and don't need this manual wiring.

Then, make `details` the `diagnosticKey` such that it will generate a corresponding JSEnd error.

paweljakubas replied:
done

KtorZ commented on src/Cardano/Wallet/Kernel/Transactions.hs:
```diff
@@ -216,6 +233,10 @@ newUnsignedTransaction ActiveWallet{..} options accountId payees = runExceptT $
 
+
+    _ <- withExceptT NewTransactionNotEnoughUtxoFragmentation $ exceptT $
+           determineCorrectUtxoFragmentation payees availableUtxo
```
I'll suggest to change the signature of `determineCorrectUxoFragmentation` to `checkUtxoFragmentation :: NonEmpty (Address, Coin) -> Utxo -> Either NumberOfMissingUtxos ()` 

There's no need to return something here and the check is really just about stopping the computation or, moving forward.

paweljakubas replied:
done

KtorZ commented on test/integration/Test/Integration/Scenario/Transactions.hs:
```diff
@@ -4,12 +4,116 @@ module Test.Integration.Scenario.Transactions
+                      noSpendingPassword
+                      NormalAssurance
+                      "NewWallet1"
+                      CreateWallet
```
Please use the fixture for that. Fixture is used to create an initial wallet without the hassle of generating all the bits yourself. Wallets come default accounts `defaultAccountId` that you can use in the scenario. So basically, the lines above can just be replaced with:

```hs
fixture01 <- setup $ defaultSetup & name .~ "newWallet1"
```

paweljakubas replied:
applied. now it is much easier to read and all imp delegated to DSL

KtorZ commented on test/integration/Test/Integration/Scenario/Transactions.hs:
```diff
@@ -4,12 +4,116 @@ module Test.Integration.Scenario.Transactions
+                      "NewWallet2"
+                      CreateWallet
+        account2 <- successfulRequest $ Client.getAccounts
+                    $- wallet2 ^. walletId
```
Same as above :+1:

paweljakubas replied:
:+1:

KtorZ commented on test/integration/Test/Integration/Scenario/Transactions.hs:
```diff
@@ -4,12 +4,116 @@ module Test.Integration.Scenario.Transactions
+                $ zipWith
+                Client.PaymentDistribution
+                recepientWalAddresses
+                (map (\coin -> Client.WalletCoin $ mkCoin coin) [10,11])
```
We really want to keep the complexity and the internal logic outside of the scenario. The `DSL` already offers:

```hs
defaultDistribution
    :: HasType (NonEmpty Address) s
    => Word64
    -> s
    -> NonEmpty PaymentDistribution
```

which isn't enough in your case, so I'd suggest to add a new:

```hs
customDistribution
    :: NonEmpty (Address, Word64)
    -> NonEmpty PaymentDistribution
```

paweljakubas replied:
done

KtorZ commented on test/integration/Test/Integration/Scenario/Transactions.hs:
```diff
@@ -4,12 +4,116 @@ module Test.Integration.Scenario.Transactions
+
+        verify response
+            [ expectTxInHistoryOf (fixture ^. wallet)
+            , expectTxStatusEventually [Creating, Applying, InNewestBlocks, Persisted]
```
I'd rather wait for `InNewestBlocks` here, to make sure the transaction is actually indeed accepted by the nodes.

paweljakubas replied:
indeed -> done

KtorZ commented on src/Cardano/Wallet/API/V1/Types.hs:
```diff
@@ -2410,6 +2412,26 @@ instance FromJSON ErrNotEnoughMoney where
+        if missingUtxos == 1 then
+            bprint ("Missing "%build%" utxo to accommodate all outputs of the transaction") missingUtxos
+        else
+            bprint ("Missing "%build%" utxos to accommodate all outputs of the transaction") missingUtxos
```
Nice effort :smile: 
Pro tip: an `(s)` works marvelously for this kind of things:

```hs
             bprint ("Missing "%build%" utxo(s) to accommodate all outputs of the transaction") missingUtxos 
```

And, in general, as a backend, the accuracy of your grammar on edge-cases is really your last concern ^^"

paweljakubas replied:
yeah, a little bit overstretched here :-) corrected

KtorZ commented on test/unit/Servant/JsendCompliance.hs:
```diff
@@ -20,7 +20,7 @@ import           Servant.Swagger.Internal.Test (props)
                      Every, Nub, TMap)
 import           Test.QuickCheck (Arbitrary (..), Property, counterexample,
-                     (.&&.), (===))
+                     (===))
```
Minor remark: when you fix something like this (which has nothing to do with our PR per se), make it a separate commit. Such that, the history is clearer to understand and, commits can be cherry-picked / reverted easily when they need to. 

Thanks for fixing the warning though :+1:

KtorZ replied:
Also, this happens because `-WError` isn't on in `cardano-wallet.cabal` for the unit tests. So I'd rather turn it on as well while we're at it :)

paweljakubas replied:
reverted the change, and included in another commit (together with switching on of -Werror flag for unit tests)

KtorZ commented on test/integration/Test/Integration/Scenario/Transactions.hs:
```diff
@@ -49,7 +104,7 @@ spec = do
 
         verify response
-            [ expectWalletError (NotEnoughMoney (ErrAvailableBalanceIsInsufficient 0))
+            [ expectWalletError (UtxoNotEnoughFragmented (Client.ErrUtxoNotEnoughFragmented 1 Client.msgUtxoNotEnoughFragmented))
```
:+1: Note that, this is a small "breaking" change and we should notify the frontend folks about this!

paweljakubas replied:
Will inform Nicola team about the change

KtorZ commented on src/Cardano/Wallet/API/V1/Types.hs:
```diff
@@ -2657,3 +2689,5 @@ instance HasDiagnostic WalletError where
         RequestThrottled{} ->
             "microsecondsUntilRetry"
+        UtxoNotEnoughFragmented{} ->
+            "msg"
```
Note, by using "msg" as a diagnostic key, we'll end up with an error like:

```json
{    
    "status": "error", 
    "diagnostic": {        
        "msg": {            
            "msg": "...",            
            "missingUtxos": 14        
        }    
    },    
    "message": "UtxoNotEnoughFragmented"
}
```

So that's quite a lot of 'msg / messages' ... May I recommand to change the diagnostic key to `details` and maybe also, to change the `theMsg` by `theHelp` in the `ErrUtxoNotEnoughFragmented` type ? This will generate something more like:

```json
{    
    "status": "error", 
    "diagnostic": {        
        "details": {            
            "help": "...",            
            "missingUtxos": 14        
        }    
    },    
    "message": "UtxoNotEnoughFragmented"
}
```

paweljakubas replied:
done

KtorZ commented on test/integration/Test/Integration/Framework/DSL.hs:
```diff
@@ -265,6 +269,18 @@ defaultDistribution
+    -> NonEmpty PaymentDistribution
+customDistribution accounts coins =
+    let recepientWalAddresses = map (\acc -> acc ^. walAddresses)
+                                $ concatMap (\acc -> acc ^. addresses)
```
NOTE: `\x -> x ^. myLens === view myLens`, favor functions over operator when passing functions around :+1:

KtorZ commented on test/integration/Test/Integration/Framework/DSL.hs:
```diff
@@ -265,6 +269,18 @@ defaultDistribution
+                                $ concatMap (\acc -> acc ^. addresses)
+                                $ accounts
+    in fromList $
+       zipWith PaymentDistribution recepientWalAddresses (map (\coin -> WalletCoin $ mkCoin coin) coins)
```
This may lead to unexpected results if:

- The `accounts` and `coins` list don't have the same number of elements
- The `accounts` have more than one address 

I'd recommend to either check that as invariant and fail hard if the invariant aren't met, or (better IMO), require the caller to provide the a `(Account, Word64)]` himself and only take the first address of each account -- failing if there's none maybe. Or asking directly for `[(Address, Word64)]`

paweljakubas replied:
done

KtorZ commented:
Nice work @paweljakubas! Thanks. Don't forget to put your ticket in QA after giving a few pointers to @piotr-iohk!

rvl committed:
Add functions for serializing keys as text

To store XPrv and XPub in a database column or text file, a
hex-encoded string is best.

KtorZ reviewed (commented):
:+1: 
Looks good, let's add unit tests corresponding to these too.

paweljakubas committed:
Add roundtrip tests

address review remarks

KtorZ commented on lib/core/test/unit/Cardano/Wallet/Primitive/AddressDerivationSpec.hs:
```diff
@@ -220,6 +234,22 @@ prop_roundtripEnumIndexSoft :: Index 'Soft 'AddressK -> Property
+    -> Property
+prop_roundtripXPriv xpriv = do
+    let (Right xpriv') = (deserializeXPrv . serializeXPrv) xpriv
+    xpriv === xpriv'
```
This should rather be written:

```hs
xpriv' = (deserializeXPrv . serializeXPrv) xpriv 
Right xpriv === xpriv'
```

With the latter version, in case the property fails, we get a nice error output with a counter example from Quickcheck. With the former, it blows up because of an invalid pattern-match :grimacing:

paweljakubas replied:
corrected

KtorZ commented on lib/core/test/unit/Cardano/Wallet/Primitive/AddressDerivationSpec.hs:
```diff
@@ -313,3 +343,35 @@ instance {-# OVERLAPS #-} Arbitrary (Passphrase "encryption") where
+genRootKeys :: Gen (Key 'RootK XPrv)
+genRootKeys = do
+    (s, g, e) <- (,,)
+        <$> genPassphrase @"seed" (0, 32)
```
Minor remark but the seed should be at least 16-bit according to `cardano-crypto`:

```hs
-- | Generate a new XPrv from an entropy seed
--
-- The seed should be at least 16 bytes, although it is not enforced
--
-- The passphrase encrypt the secret key in memory
generateNew :: (ByteArrayAccess keyPassPhrase, ByteArrayAccess generationPassPhrase, ByteArrayAccess seed)
```

We don't actually have any comment on the matter in our code either. I'll put a point on the blackboard.

paweljakubas replied:
corrected
